src/v1.rs

```diff
@@ -162,25 +149,36 @@ impl Party {
             return Err(DkgError::BadPublicShares(bad_ids));
         }
 
+        let mut missing_shares = Vec::new();
+        for i in public_shares.keys() {
+            if private_shares.get(i).is_none() {
+                missing_shares.push((self.id, *i));
+            }
+        }
+        if !missing_shares.is_empty() {
+            return Err(DkgError::MissingPrivateShares(missing_shares));
+        }
+
         // let's optimize for the case where all shares are good, and test them as a batch
 
         // building a vector of scalars and points from public poly evaluations and expected values takes too much memory
         // instead make an object which implements p256k1 MultiMult trait, using the existing powers of x and shares
-        let mut check_shares = CheckPrivateShares::new(self.id(), &shares, polys.clone());
+        let mut check_shares =
+            CheckPrivateShares::new(self.id(), &private_shares, public_shares.clone());
 
         // if the batch verify fails then check them one by one and find the bad ones
         if Point::multimult_trait(&mut check_shares)? != Point::zero() {
             let mut bad_shares = Vec::new();
-            for (i, s) in shares.iter() {
-                let comm = &polys[i];
+            for (i, s) in private_shares.iter() {
+                let comm = &public_shares[i];
```

Comment:
Done.  It should be safe to index, since we checked above that the maps contain the correct keys.  But switching to get will future proof this against someone making changes without realizing the consequences.